[SYCL]Use llvm-link's only-needed option to link device libs

jinge90 opened the pull request:
Signed-off-by: gejin <ge.jin@intel.com>

This PR aims to solve the same problem in a different way from https://github.com/intel/llvm/pull/2723 which is a better way.
We will do 2-step link:
1. normal llvm-link for all user LLVM bc to generate a "full" user.bc
2. llvm-link -only-needed to link "full" user.bc with device libraries.
In step, we expect only functions really required by user code will be linked in.
We create the PR to run pre-ci tests and CPU performance tests.

jinge90 commented:
/summary:run

jinge90 commented:
Hi, @bader 
This PR is based on your suggestion "option 2", we split the device link into 2 steps:
1. normal llvm-link for all user's bc
2. llvm-link only needed for devicelib bc.
The code change is little, I ran full test suite and all passed. Also verified some perf benchmarks on CPU device, the perf improvement is as expected. Based on the IR dumped, no "double" functions were introduced into final device image if user code didn't include any double function, I will try on real platforms without double support later.
I will have a look into your "option 1" later.
Thanks very much.

bader commented:
@jinge90, thanks. Let me know the results of your investigation. I suggest closing #2723 - this solution is preferable in my opinion.

jinge90 commented:
> @jinge90, thanks. Let me know the results of your investigation. I suggest closing #2723 - this solution is preferable in my opinion.

Yes, I will close #2723, no matter option 1 works or not, this PR is much better than #2723.
Thanks very much.

mdtoguchi reviewed (commented):
Could a test be added to verify when -only-needed is added?

jinge90 commented:
> Could a test be added to verify when -only-needed is added?

Hi, @mdtoguchi 
We need  some more investigation before making final decisions to optimize devicelib default. Once we decide to use this PR, I will add driver test to verify "-only-needed".
Thanks very much.

jinge90 commented:
> Could a test be added to verify when -only-needed is added?

Hi, @mdtoguchi 
I have updated clang driver test sycl-device-lib and sycl-device-lib-win to add check for "llvm-link -only-needed" and I have updated both tests to adapt to latest compiler and device libraries(e.g. libsycl-glibc/msvc.o have been renamed to libsycl-crt.o).
Thanks very much.

jinge90 commented:
/summary:run

jinge90 commented:
Hi, @bader @vzakhari @mdtoguchi 
Could you help to review this patch?
Thanks very much.

bader commented on clang/test/Driver/sycl-device-lib-win.cpp:
```diff
@@ -1,7 +1,7 @@
 /// Perform several driver tests for SYCL device libraries on Windows
 ///
-// REQUIRES: clang-driver, windows
+// UNSUPPORTED: system-linux
```
Deleted code seems to express the limitation more precisely. Is it supposed to work on Darwin, Fuchsia or FreeBSD?

jinge90 replied:
Hi, @bader 
You are correct, this lit test should be only for windows.
Thanks very much.

bader commented on clang/lib/Driver/ToolChains/SYCL.cpp:
```diff
@@ -101,6 +100,18 @@ const char *SYCL::Linker::constructLLVMLinkCommand(Compilation &C,
     StringRef SubArchName, StringRef OutputFilePrefix,
     const InputInfoList &InputFiles) const {
   ArgStringList CmdArgs;
+
+  bool LinkSYCLDeviceLibs = false;
+  for (const auto &II : InputFiles) {
+    StringRef InputFilename =
+        llvm::sys::path::filename(StringRef(II.getFilename()));
+    StringRef InputSuffix = ".o";
+    if (InputFilename.startswith("libsycl-") &&
+        InputFilename.endswith(InputSuffix)) {
+      LinkSYCLDeviceLibs = true;
+      break;
+    }
+  }
```
I think it's better to move this code under if statement at line 120 and replace for-loop with standard library function call e.g. `std::find_if`.

bader commented on clang/lib/Driver/ToolChains/SYCL.cpp:
```diff
@@ -100,24 +100,24 @@ const char *SYCL::Linker::constructLLVMLinkCommand(Compilation &C,
+          if (InputFilename.startswith("libsycl-") &&
+              InputFilename.endswith(".o"))
+            return true;
+          else
+            return false;
```
```suggestion
          return false;
```

bader commented on clang/lib/Driver/ToolChains/SYCL.cpp:
```diff
@@ -100,24 +100,24 @@ const char *SYCL::Linker::constructLLVMLinkCommand(Compilation &C,
   // an actual object/archive.  Take that list and pass those to the linker
   // instead of the original object.
   if (JA.isDeviceOffloading(Action::OFK_SYCL)) {
+    bool LinkSYCLDeviceLibs = false;
```
```suggestion
```

bader commented on clang/lib/Driver/ToolChains/SYCL.cpp:
```diff
@@ -100,24 +100,24 @@ const char *SYCL::Linker::constructLLVMLinkCommand(Compilation &C,
+          else
+            return false;
+        });
+    LinkSYCLDeviceLibs = (SYCLDeviceLibIter != InputFiles.end());
```
```suggestion
    bool LinkSYCLDeviceLibs = (SYCLDeviceLibIter != InputFiles.end());
```

mdtoguchi commented on clang/test/Driver/sycl-device-lib-win.cpp:
```diff
@@ -91,3 +91,23 @@
+/// test llvm-link behavior for fno-sycl-device-lib
+// RUN: %clangxx -fsycl -fno-sycl-device-lib=all %s -### 2>&1 \
+// RUN:   | FileCheck %s -check-prefix=SYCL_LLVM_LINK_NO_DEVICE_LIB
+// SYCL_LLVM_LINK_NO_DEVICE_LIB: llvm-link{{.*}}  "{{.*}}.bc" "-o" "{{.*}}.bc" "--suppress-warnings"
```
I would expect that `-only-needed` is not passed in this case - can a negative check be added?

jinge90 commented on clang/test/Driver/sycl-device-lib.cpp:
```diff
@@ -91,3 +92,24 @@
+/// test llvm-link behavior for fno-sycl-device-lib
+// RUN: %clangxx -fsycl -fno-sycl-device-lib=all %s -### 2>&1 \
+// RUN:   | FileCheck %s -check-prefix=SYCL_LLVM_LINK_NO_DEVICE_LIB
+// SYCL_LLVM_LINK_NO_DEVICE_LIB: clang{{.*}} "-cc1" {{.*}} "-fsycl-is-device"
```
@mdtoguchi 
I have updated the test case to check if "only-needed" is excluded when "-fno-sycl-device-lib=all" is used.
Thanks very much.

mdtoguchi reviewed (approved):
LGTM

bader commented:
/summary:run

sndmitriev commented on clang/lib/Driver/ToolChains/SYCL.cpp:
```diff
@@ -107,8 +106,20 @@ const char *SYCL::Linker::constructLLVMLinkCommand(Compilation &C,
+          if (InputFilename.startswith("libsycl-") &&
+              InputFilename.endswith(".o"))
+            return true;
+          return false;
```
Using file name for deciding if `--only-needed` option is required looks very fragile.

jinge90 replied:
Hi, @sndmitriev 
I agree, we need some enhancement here .
